src/EditorFeatures/CSharp/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,428 @@
+            if (blockCommentSelectionHelper.IsLocationCommented(selectedSpan.Start) && !blockCommentSelectionHelper.DoesBeginWithBlockComment(commentInfo))
+            {
+                InsertText(textChanges, selectedSpan.Start, commentInfo.BlockCommentEndString);
+                InsertText(textChanges, selectedSpan.Start, commentInfo.BlockCommentStartString);
+                // Shrink the tracking so the previous comment start marker is not included in selection.
+                amountToAddToStart = commentInfo.BlockCommentEndString.Length;
+            }
+
+            // If the end is commented (and not a comment marker), close the current comment and open a new one.
+            if (blockCommentSelectionHelper.IsLocationCommented(selectedSpan.End) && !blockCommentSelectionHelper.DoesEndWithBlockComment(commentInfo))
+            {
+                InsertText(textChanges, selectedSpan.End, commentInfo.BlockCommentEndString);
+                InsertText(textChanges, selectedSpan.End, commentInfo.BlockCommentStartString);
+                // Shrink the tracking span so the next comment start marker is not included in selection.
+                amountToAddToEnd = -commentInfo.BlockCommentStartString.Length;
+            }
+
+            trackingSpans.Add(blockCommentSelectionHelper.GetTrackingSpan(selectedSpan, spanTrackingMode, Operation.Comment, amountToAddToStart, amountToAddToEnd));
+        }
+
+        private static void AddBlockComment(CommentSelectionInfo commentInfo, Span span, List<TextChange> textChanges)
+        {
+            InsertText(textChanges, span.Start, commentInfo.BlockCommentStartString);
+            InsertText(textChanges, span.End, commentInfo.BlockCommentEndString);
+        }
+
+        private static void DeleteBlockComment(BlockCommentSelectionHelper blockCommentSelectionHelper, Span spanToRemove, List<TextChange> textChanges,
+            CommentSelectionInfo commentInfo)
+        {
+            DeleteText(textChanges, new TextSpan(spanToRemove.Start, commentInfo.BlockCommentStartString.Length));
+
+            var blockCommentMarkerPosition = spanToRemove.End - commentInfo.BlockCommentEndString.Length;
+            // Sometimes the block comment will be missing a close marker.
+            if (Equals(blockCommentSelectionHelper.GetSubstringFromText(blockCommentMarkerPosition, commentInfo.BlockCommentEndString.Length), commentInfo.BlockCommentEndString))
+            {
+                DeleteText(textChanges, new TextSpan(blockCommentMarkerPosition, commentInfo.BlockCommentEndString.Length));
+            }
+        }
+
+        private static IEnumerable<Span> GetDescendentBlockCommentSpansFromRoot(SyntaxNode root)
+        {
+            return root.DescendantTrivia()
+                .Where(trivia => trivia.IsKind(SyntaxKind.MultiLineCommentTrivia) || trivia.IsKind(SyntaxKind.MultiLineDocumentationCommentTrivia))
+                .Select(blockCommentTrivia => blockCommentTrivia.Span.ToSpan());
```

Comment:
avoid this.  It will prevent TypeScript and other languages from properly working with roslyn with this command.  That violates a core design of our handlers and how we plug in otehr languages.  Specifically, we have the handler, but we also have a Service that *any* Roslyn language can implement.  If they implement the service, then things will light up.  This is beneficial so that all these languages do not have to reimplement the same core logic for all features.

Reply:
I see, so the handler could be called into by non-C# / VB.  

If for the moment we only wanted C#, would a resolution be to provide a service for C# that say returns the commented spans (using syntax tree or w.e else).  Then depend on that service in the handler to add /remove as needed?

Reply:
> If for the moment we only wanted C#

In general, that's not a good thing to want.  It usually causes the feature to become hyperspecialized without a good breakup of responsibilities.  By supporting multi-lang, we generally create better features that are more cleanly designed and implemented.

--

However, say we wanted to just have something quick-and-dirty that we could try out *but which would not be checked into master*, then yes, i would have an IToggleCommentService that C# would implement.  It would collect whatever data was necessary, and would pass it up to the handler to deal with.

That said it's unclear to me why we'd need such a service.  Comment/Uncomment doesn't really need it.  It can figure it out from the document, just based on hte info from the ICommentSelectionService.  Is the info there insufficient?  Are you worried about it potentially being a bit wonky since it's textual?

> Then depend on that service in the handler to add /remove as needed?

In general, our services don't try to "mutate" anything.  i.e. they don't add/remove.  Instead, they just return data that describes how their language sees the world, and it's the responsibility of teh feature (i.e. the command handler here) to actually mutate the world.

Reply:
I'd prefer we just stick with ICommentSelectionService.

If there's a super important need for new information that that doesn't provide, tehn you can def add a new service.  But i'd like to understand exactly what data is returned back.  Thanks!

Reply:
Sorry, what I meant to say is not a service specifically to toggle comments, but just to retrieve existing comments.  I'm mainly using the syntax tree here to retrieve commented spans (so I don't have to deal with weird issues like ```string s = "/*";``` by just parsing comment markers).

The handler would still have the logic to handle checking the selected spans and removing or adding comments, but wouldn't depend on C# concepts like the syntax tree.

It could be added functionality to the existing ICommentSelectionService as well, I don't mind.

Reply:
@dibarbet That would certainly work.  I would prefer:

1. try to do this on the existing ICommentSelectionService, using just the existing properties.
2. if that's not possible, do what you mentioned above.  i.e. provide an API that will enumerate the relevant block-comment spans.

Reply:
@CyrusNajmabadi discussed it around here a bit today
The suggestion is to add a new method to the existing ICommentSelectionService for toggling block comment (take in doc, return out doc).

If we just did return the comment spans and leave the logic in the handler, then they get tied into whatever toggle functionality we define (e.g. adding sequential comments or w.e. else)

Reply:
I'm personally opposed to that because it breaks symettry with how comment/uncomment works.  

> then they get tied into whatever toggle functionality we define (e.g. adding sequential comments or w.e. else)

THat doesn't seem problematic to me.  I think we could define that.  And we'd have VB/C#/TS/JS/F# to validate our approach (like we did with the normal comment/uncomment work).  Is there a belief that we could not do this?  It wasn't particularly costly or hard the first time around.  SO i'd prefer we maintain this approach for "toggle comment" as well.

Reply:
since we don't know how all language want to handle block comments, I believe document in document out is simplest thing we can do for now. and if, I mean if any other language decide to support it such as typescript, F# and xaml, then we can then later add something that works for everyone once we know what all wants to do.

right now, we thinking how all other language would use this is just over engineering with no concrete data backing it up.

Reply:
>  and if, I mean if any other language decide to support it such as typescript, F# and xaml, then we can then later add something that works for everyone once we know what all wants to do.

It's very simple for us to make this determination by having a 5 minutes conversation with these teams.  This is what we already did for the "CommentUncomment" feature.  It was established that we all wanted the same behavior.  Which is why we unified on a single implementation, that way you didn't have to reinvent the wheel for every language.

I am concerned we're not even trying to follow the same path forward here for something which is obviously in the exact same feature area.  

>right now, we thinking how all other language would use this is just over engineering with no concrete data backing it up.

No.  It's not overengineering.  It's soundly thinking about reducing unnecessary duplication.  Do we expect that TS/JS would want this feature?  We're smart, experienced, IDE devs and we can confidently say: yes they would want this (like they wanted comment/uncomment).

Do we expect that we could trivial learn about the behavior they want by havin a 5 minute conversation with them?  Yes.  We know this from all our other experiences with these sorts of features (including comment/uncomment).

Please just work with our partner teams.  It takes almost no effort and it can save a lot of wasted and duplicated effort.  This is why we built the roslyn extensibility model in the first place, and why we deliberately implemented features like Comment/Uncomment in this manner.  There is no reason to have to slip backward here when it takes practically *zero* effort to just do the right thing.